### PR TITLE
fix: Retry requests from provider after connection is ready, on all builds

### PR DESCRIPTION
**Pull request by jiexi**
## **Description**

Currently, requests from the dapp can hang if: 
1. they are sent before the extension is fully initialized / before the extension is ready to process requests from that specific dapp connection
2. they get swallowed do to an issue in chromium with streams on prerendered pages breaking when they become visible to the user

The inpage provider does have a retry mechanism, but currently that only works on MV3 builds. Additionally, is not a 100% reliable trigger as the underlying `metamask_chainChanged` message that it relies on is tied to MetaMaskController mem state changes, and is not fired off each time a dapp connection is established.

This PR attempts to fix request retries by:
* Making the retry message triggered by `metamask_chainChanged` work on all builds, not just MV3 builds
* Firing `metamask_chainChanged` every time a new dapp connection is established

**Open concern:**
There may be an edge case when a request with a confirmation is sent by the dapp, then received and shown to the user in the wallet, but swallowed by prerender. I believe my proposed changes in this PR would cause the confirmation to get resent even though there may still be a confirmation being shown to the user. I haven't been able to test this properly as debugging the prerender issue in general is difficult. 

This would require a dapp to send a request requiring confirmation pretty much immediately after page load, implying without user interaction.

[Open in GitHub Codespaces](https://codespaces.new/MetaMask/metamask-extension/pull/24142?quickstart=1)

## **Related issues**

See: https://github.com/MetaMask/metamask-extension/issues/11069
See: https://github.com/MetaMask/metamask-extension/issues/23329
See: https://github.com/MetaMask/metamask-extension/issues/24025

## **Manual testing steps**

NOTE: these testing steps require that you add an artificial 5s delay to the background connection handling. See the comments below to find the commit where this delay was included in this PR before being removed for merging

### Verify this fixes the `metamask_sendDomainMetadata` issue
1. Restart extension
2. Open any page
3. Quickly open the wallet popup
4. Notice that there is no favicon
5. After 5s have elapsed since the page was opened, the favicon should populate in the wallet popup

### Verify this fixes retrying messages before extension is ready
1. Open any page (restarting extension not required)
2. Quickly open console and type `await window.ethereum.request({method: 'eth_chainId'})`
3. Notice that the response doesn't come immediately
4. After 5s have elapsed since the page was opened, the response should be shown in console

### Verify this fixes retrying messages after broken prerender stream
Not easy to verify this unfortunately, you would need to break the prerender stream while a request was being made

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: app/scripts/metamask-controller.js [hunk 1]
```diff
@@ -4825,6 +4825,9 @@ export default class MetamaskController extends EventEmitter {
         log.error(err);
       }
     });
+
+    // Used to show wallet liveliness to the provider
+    this._notifyChainChange();
```
Comment (jiexi) [inline app/scripts/metamask-controller.js:4830]: note that this fires a `metamask_chainChanged` to all connections, not just the new one. Need to update the code to change this, but also it's starting to feel wrong using `metamask_chainChanged` as a liveliness check

File: app/scripts/background.js [hunk 2]
```diff
@@ -707,6 +707,9 @@ export function setupController(
     } else {
       // this is triggered when a new tab is opened, or origin(url) is changed
       if (remotePort.sender && remotePort.sender.tab && remotePort.sender.url) {
+        console.log(`delaying connection to ${remotePort.sender.url} for 5s`);
+        await new Promise((resolve) => setTimeout(resolve, 5000));
+
```
Comment (jiexi) [inline app/scripts/background.js:712]: this is here to make testing easier. It delays the handling of the dapp responses over the port for 5s.

Comment (jiexi) [reply app/scripts/background.js:712]: TODO: Remove this before approval, of course

**Comment by jiexi** [comment]
Additionally, we can remove the prerender case entirely by disabling it from the extension. This will hurt perceived browser responsiveness in certain cases. Other extensions such as uBlock do this already. I myself have prerender disabled due to uBlock and it has not noticeably impacted my browsing experience

**Review by jpuri — DISMISSED** [review]

File: app/scripts/metamask-controller.js (hunk 1 above)
Comment (jiexi) [reply app/scripts/metamask-controller.js]: notify only new connection here https://github.com/MetaMask/metamask-extension/pull/24142/commits/f90b220061cd7252f2376b827ee9759dee236ae6

**Comment by jiexi** [comment]
Following change in web3-stream-provider needed https://github.com/MetaMask/web3-stream-provider/pull/16

File: app/scripts/background.js (hunk 2 above)
Comment (jiexi) [reply app/scripts/background.js]: This has been removed, but must be added back locally if you are testing manually

**Comment by adonesky1** [comment]
I'm not sure what this is about but I'm seeing some strange inexplicable inconsistency with favicons:
https://github.com/MetaMask/metamask-extension/assets/34557516/87cf7091-fd41-4c42-ab8c-d1c9082a0a6e

**Review by adonesky1 — APPROVED** [review]
I'm not seeing the issue with the favicon anymore. I've tested all scenarios (with and without the added delay) and they all look good. Code looks good to me!

**Review by shanejonas — APPROVED** [review]

**Review by Gudahtt — APPROVED** [review]
LGTM!

**Pull request merged**